x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -291,20 +291,46 @@ public Automaton allowedActionsMatcher(String index) {
                     if (actionCheck || bwcMappingActionCheck) {
                         // propagate DLS and FLS permissions over the concrete indices
                         for (String index : concreteIndices) {
-                            Set<FieldPermissions> fieldPermissions = fieldPermissionsByIndex.computeIfAbsent(index, (k) -> new HashSet<>());
-                            fieldPermissionsByIndex.put(indexOrAlias, fieldPermissions);
-                            fieldPermissions.add(group.getFieldPermissions());
-                            DocumentLevelPermissions permissions =
-                                    roleQueriesByIndex.computeIfAbsent(index, (k) -> new DocumentLevelPermissions());
-                            roleQueriesByIndex.putIfAbsent(indexOrAlias, permissions);
+
+                            Set<FieldPermissions> fieldPermissions = fieldPermissionsByIndex.get(index);
+                            if (fieldPermissions == null) {
+                                // Most indices rely on the default (empty) field permissions object, so we optimize for that case
+                                // Using an immutable single item set is significantly faster because it avoids any of the hashing
+                                // and backing set creation.
+                                fieldPermissions = Set.of(group.getFieldPermissions());
+                                fieldPermissionsByIndex.put(index, fieldPermissions);
+                            } else if (fieldPermissions.size() == 1) {
+                                FieldPermissions fp = group.getFieldPermissions();
+                                if (fieldPermissions.contains(fp) == false) {
+                                    // This index doesn't have a single field permissions object, replace the singleton with a real Set
+                                    final FieldPermissions existing = fieldPermissions.iterator().next();
+                                    fieldPermissions = new HashSet<>();
+                                    fieldPermissions.add(existing);
+                                    fieldPermissions.add(fp);
+                                    fieldPermissionsByIndex.put(index, fieldPermissions);
+                                }
+                            } else {
+                                fieldPermissions.add(group.getFieldPermissions());
+                            }
+
+                            DocumentLevelPermissions docPermissions;
                             if (group.hasQuery()) {
-                                permissions.addAll(group.getQuery());
+                                docPermissions = roleQueriesByIndex.computeIfAbsent(index, (k) -> new DocumentLevelPermissions());
+                                docPermissions.addAll(group.getQuery());
```

Comment:
By the same reasoning for FieldPermissions, should we also get rid of the usage of `computeIfAbsent`? 
```suggestion
                                docPermissions = roleQueriesByIndex.get(index);
                                if (docPermissions == null) {
                                    docPermissions = new DocumentLevelPermissions();
                                    roleQueriesByIndex.put(index, docPermissions);
                                }
                                docPermissions.addAll(group.getQuery());
```

Reply:
Generally speaking `HashMap.computeIfAbsent` is more efficient than a `get` and `put`

The reason to drop it from the FLS case is that we want to have specific cases for different circumstances, and it's more readable to do that with a `get` and `null` check.

Technically the FLS case would be slightly more efficient as:
```
   // Most indices rely on the default (empty) field permissions object, so we optimize for that case
   // Using an immutable single item set is significantly faster because it avoids any of the hashing
   // and backing set creation.
   Set<FieldPermissions> fieldPermissions = fieldPermissionsByIndex.computeIfAbsent(index, k-> Set.of(group.getFieldPermissions()));
   if (fieldPermissions != null) {
       if (fieldPermissions.size() == 1) {
           FieldPermissions fp = group.getFieldPermissions();
           if (fieldPermissions.contains(fp) == false) {
               // This index doesn't have a single field permissions object, replace the singleton with a real Set
               final FieldPermissions existing = fieldPermissions.iterator().next();
               fieldPermissions = new HashSet<>();
               fieldPermissions.add(existing);
               fieldPermissions.add(fp);
               fieldPermissionsByIndex.put(index, fieldPermissions);
           }
        } else {
           fieldPermissions.add(group.getFieldPermissions());
       }
   }
```

But it's much less readable for a very minor performance gain.

Reply:
OK. Thanks for the explanation. I thought the reason was because `get` + `put` is more efficient. In this case, should we replace them with `fieldPermissionsByIndex.compute(index, (k, fieldPermissions) -> { ... })`?

It should be pretty much equally readable, basically just move your changes into the `{ ... }`, and maybe(?) slightly more efficient?

Reply:
I've switched to `.compute'.